firmware/stellar.c

```diff
@@ -0,0 +1,876 @@
+ * type (which is always ED25519)
+ */
+void stellar_xdr_read_address(uint8_t *out_addr_bytes, uint8_t *bytestream, uint32_t *offset)
+{
+    // Read key type (always 0, so ignored)
+    uint32_t key_type = stellar_xdr_read_uint32(bytestream, offset);
+    if (key_type != 0) {
+        fsm_sendFailure(FailureType_Failure_UnexpectedMessage, _("Unsupported address type"));
+        layoutHome();
+        return;
+    }
+
+    // Read next 32 bytes
+    memcpy(out_addr_bytes, bytestream + *offset, 32);
+    *offset += 32;
+}
+
+/*
+ * Copies the string in bytestream to out_str
+ * XDR strings are:
+ *  length (4 bytes)
+ *  string (length bytes)
+ */
+void stellar_xdr_read_string(uint8_t *out_str, uint8_t *bytestream, uint32_t *offset)
+{
+    // First 4 bytes are string length
+    uint32_t strlen = stellar_xdr_read_uint32(bytestream, offset);
+
+    // Read next strlen bytes
+    memcpy(out_str, bytestream + *offset, strlen);
+    *offset += strlen;
+}
+
+uint32_t stellar_getXdrOffset()
+{
+    return stellar_activeTx.xdr_offset;
+}
+
+uint8_t stellar_allOperationsConfirmed()
+{
+    return stellar_activeTx.confirmed_operations == stellar_activeTx.num_operations;
+}
+
+StellarTransaction *stellar_getActiveTx()
+{
+    return &stellar_activeTx;
+}
+
+/*
+ * Calculates and sets the signature for the active transaction
+ */
+void stellar_getSignatureForActiveTx(uint8_t *out_signature)
+{
+    HDNode *node = stellar_deriveNode(stellar_activeTx.account_index);
+
+    // Signature is the ed25519 detached signature of the sha256 of all the bytes
+    // that have been read so far
+    uint8_t to_sign[32];
+    sha256_Final(&(stellar_activeTx.sha256_ctx), to_sign);
+
+    uint8_t signature[64];
+    ed25519_sign(to_sign, sizeof(to_sign), node->private_key, node->public_key + 1, signature);
+
+    memcpy(out_signature, signature, sizeof(signature));
+}
+
+/*
+ * Returns number (representing stroops) formatted as XLM
+ * For example, if number has value 1000000000 then it will be returned as "100.0"
+ */
+void stellar_format_stroops(uint64_t number, char *out, size_t outlen)
+{
+    bignum256 bn_number;
+    bn_read_uint64(number, &bn_number);
+    bn_format(&bn_number, NULL, NULL, 7, 0, false, out, outlen);
+}
+
+/*
+ * Returns a uint32 formatted as a string
+ */
+void stellar_format_uint32(uint64_t number, char *out, size_t outlen)
+{
+    bignum256 bn_number;
+    bn_read_uint32(number, &bn_number);
+    bn_format(&bn_number, NULL, NULL, 0, 0, false, out, outlen);
+}
+
+/*
+ * Returns a uint64 formatted as a string
+ */
+void stellar_format_uint64(uint64_t number, char *out, size_t outlen)
+{
+    bignum256 bn_number;
+    bn_read_uint64(number, &bn_number);
+    bn_format(&bn_number, NULL, NULL, 0, 0, false, out, outlen);
+}
+
+/*
+ * Breaks a 56 character address into 3 lines of lengths 16, 20, 20
+ * This is to allow a small label to be prepended to the first line
+ */
+const char **stellar_lineBreakAddress(uint8_t *addrbytes)
```

Comment:
Is there any reason not to use `split_message` from `layout2.c`?

Reply:
The 56-character Stellar addresses are a bit awkward since they can be displayed in 3 lines but don't wrap to an even number of characters. I went with the 16/20/20 split since it allows a little more flexibility in the UI.

For example, I can insert a small label like "to" when rendering the payment screen:

```
To: GDRXE2BQUC3AZNPV
FSCEZ76NJ3WWL25FYFK6
RGZGIEKWE4SOOHSUJUJ6
```